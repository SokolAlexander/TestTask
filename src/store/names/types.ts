export enum REQUEST_STATUS {
  'IDLE',
  'PENDING',
  'SUCCESS',
  'FAILURE',
}

export type Gender = 'male' | 'female' | '';

export type NamesState = {
  request: {
    status: REQUEST_STATUS;
    error: string | null;
  };
  name: string;
  gender: Gender;
  stats: {
    countFemale: number;
    history: {
      [name: string]: number;
    };
  };
};

export type GenderData = {
  name: string;
  gender: Gender;
  probability: number;
  count: number;
};

export interface BaseAction {
  type: string;
}

export interface GetGenderSuccess extends BaseAction {
  payload: GenderData;
}
export interface GetGenderFailure extends BaseAction {
  payload: string;
}

export type NamesActions = BaseAction | GetGenderFailure | GetGenderSuccess;
