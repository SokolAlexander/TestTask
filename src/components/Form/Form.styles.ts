import {StyleSheet} from 'react-native';
import {colors} from '../../utils/colors';

export const formStyles = StyleSheet.create({
  infoWrapper: {
    alignItems: 'center',
  },
  infoText: {
    fontSize: 24,
    color: colors.black,
    fontWeight: 'bold',
  },
  input: {
    backgroundColor: colors.white,
    borderWidth: 1,
    borderColor: colors.darkGrey,
    marginVertical: 10,
    paddingVertical: 10,
    paddingHorizontal: 15,
  },
  error: {
    color: colors.red,
  },
  statusWrapper: {
    alignItems: 'center',
    minHeight: 30,
  },
  button: {
    backgroundColor: colors.white,
    borderColor: colors.darkGrey,
    borderWidth: 2,
    padding: 15,
    alignItems: 'center',
    justifyContent: 'center',
    marginHorizontal: 25,
    marginTop: 10,
  },
  buttonDisabled: {backgroundColor: colors.lightGrey, borderColor: colors.grey},
  btnText: {
    fontSize: 18,
    color: colors.darkGrey,
  },
});
