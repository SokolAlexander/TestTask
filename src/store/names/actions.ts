import {Dispatch} from 'redux';
import {API_URL} from '../../utils/constants';
import {
  GenderData,
  GetGenderFailure,
  BaseAction,
  GetGenderSuccess,
} from './types';

export const GET_GENDER_REQUEST = 'NAMES::GET_GENDER_REQUEST';
export const GET_GENDER_SUCCESS = 'NAMES::GET_GENDER_SUCCESS';
export const GET_GENDER_FAILURE = 'NAMES::GET_GENDER_FAILURE';

export const getGenderRequest = (): BaseAction => ({
  type: GET_GENDER_REQUEST,
});

export const getGenderSuccess = (genderData: GenderData): GetGenderSuccess => ({
  type: GET_GENDER_SUCCESS,
  payload: genderData,
});

export const getGenderFailure = (errorMsg: string): GetGenderFailure => ({
  type: GET_GENDER_FAILURE,
  payload: errorMsg,
});

export const getGender = (name: string) => async (dispatch: Dispatch) => {
  try {
    dispatch(getGenderRequest());

    const response = await fetch(`${API_URL}${name}`);
    const result = await response.json();

    if (!result.gender) {
      throw new Error(result.error || 'Nothing found');
    }

    dispatch(getGenderSuccess(result));
  } catch (e: any) {
    dispatch(getGenderFailure(e.message));
  }
};
