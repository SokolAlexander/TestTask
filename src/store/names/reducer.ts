import {
  GET_GENDER_FAILURE,
  GET_GENDER_REQUEST,
  GET_GENDER_SUCCESS,
} from './actions';
import {
  GetGenderFailure,
  GetGenderSuccess,
  NamesActions,
  NamesState,
  REQUEST_STATUS,
} from './types';

const initialState: NamesState = {
  request: {
    status: REQUEST_STATUS.IDLE,
    error: null,
  },
  name: '',
  gender: '',
  stats: {
    countFemale: 0,
    history: {},
  },
};

export const namesReducer = (
  state = initialState,
  action: NamesActions,
): NamesState => {
  switch (action.type) {
    case GET_GENDER_REQUEST:
      return {
        ...state,
        request: {
          status: REQUEST_STATUS.PENDING,
          error: null,
        },
      };
    case GET_GENDER_FAILURE:
      return {
        ...state,
        request: {
          status: REQUEST_STATUS.FAILURE,
          error: (action as GetGenderFailure).payload,
        },
      };
    case GET_GENDER_SUCCESS: {
      const {
        payload: {name, gender},
      } = action as GetGenderSuccess;
      const historyCount = state.stats.history[name] || 0;
      return {
        ...state,
        request: {
          ...state.request,
          status: REQUEST_STATUS.SUCCESS,
        },
        name,
        gender,
        stats: {
          countFemale:
            gender === 'female'
              ? state.stats.countFemale + 1
              : state.stats.countFemale,
          history: {
            ...state.stats.history,
            [name]: historyCount + 1,
          },
        },
      };
    }
    default:
      return state;
  }
};
