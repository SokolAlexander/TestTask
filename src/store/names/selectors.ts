import {NamesState} from './types';

export const selectName = (state: NamesState) => state.name;
export const selectGender = (state: NamesState) => state.gender;

export const selectRequestStatus = (state: NamesState) => state.request.status;
export const selectRequestError = (state: NamesState) => state.request.error;

export const selectHistory = (state: NamesState) => state.stats.history;
export const selectCountFemale = (state: NamesState) => state.stats.countFemale;
