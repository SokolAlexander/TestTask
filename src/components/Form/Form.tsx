import React, {useState} from 'react';
import {
  ActivityIndicator,
  Text,
  TextInput,
  TouchableOpacity,
  View,
} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import {getGender} from '../../store/names/actions';
import {
  selectGender,
  selectName,
  selectRequestError,
  selectRequestStatus,
} from '../../store/names/selectors';
import {REQUEST_STATUS} from '../../store/names/types';
import {formStyles} from './Form.styles';

export const Form = () => {
  const dispatch = useDispatch();
  const name = useSelector(selectName);
  const gender = useSelector(selectGender);
  const error = useSelector(selectRequestError);
  const status = useSelector(selectRequestStatus);

  const [value, setValue] = useState('');

  const handlePress = () => {
    dispatch(getGender(value));
    setValue('');
  };

  const btnDisabled = status === REQUEST_STATUS.PENDING || !value;

  return (
    <>
      {!!name && (
        <View style={formStyles.infoWrapper}>
          <Text style={formStyles.infoText}>
            {name}: {gender}
          </Text>
        </View>
      )}
      <TextInput
        style={formStyles.input}
        value={value}
        onChangeText={setValue}
        placeholder="Enter a name"
      />
      <View style={formStyles.statusWrapper}>
        {!!error && <Text style={formStyles.error}>{error}</Text>}
        {status === REQUEST_STATUS.PENDING && (
          <ActivityIndicator color="black" />
        )}
      </View>
      <TouchableOpacity
        disabled={btnDisabled}
        style={[formStyles.button, btnDisabled && formStyles.buttonDisabled]}
        onPress={handlePress}>
        <Text style={formStyles.btnText}>Get Gender</Text>
      </TouchableOpacity>
    </>
  );
};
