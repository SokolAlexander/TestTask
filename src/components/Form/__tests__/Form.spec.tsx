import React from 'react';
import {fireEvent, render} from '@testing-library/react-native';

import {Form} from '../Form';
import {
  createTestWrapper,
  defaultMockState,
  defaultMockStore,
  makeStore,
} from '../../../utils/createTestWrapper';
import {Gender, REQUEST_STATUS} from '../../../store/names/types';
import * as actions from '../../../store/names/actions';

const DefaultTestWrapper = createTestWrapper();
jest.spyOn(actions, 'getGender');

describe('Form', () => {
  it('matches snapshot with default state', () => {
    const component = render(
      <DefaultTestWrapper>
        <Form />
      </DefaultTestWrapper>,
    );

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot with request pending state', () => {
    const mockStore = makeStore({
      ...defaultMockState,
      request: {status: REQUEST_STATUS.PENDING},
    });
    const TestWrapper = createTestWrapper(mockStore);
    const component = render(
      <TestWrapper>
        <Form />
      </TestWrapper>,
    );

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot with request error state', () => {
    const mockStore = makeStore({
      ...defaultMockState,
      request: {status: REQUEST_STATUS.FAILURE, error: 'Test error'},
    });
    const TestWrapper = createTestWrapper(mockStore);
    const component = render(
      <TestWrapper>
        <Form />
      </TestWrapper>,
    );

    expect(component).toMatchSnapshot();
  });

  it('shows name and gender if they are present', () => {
    const name = 'Name';
    const gender: Gender = 'female';
    const mockStore = makeStore({
      ...defaultMockState,
      name,
      gender,
    });
    const TestWrapper = createTestWrapper(mockStore);
    const {queryByText} = render(
      <TestWrapper>
        <Form />
      </TestWrapper>,
    );

    const text = queryByText(`${name}: ${gender}`);
    expect(text).not.toBeNull();
  });

  it('disables button if there is no value in input', () => {
    const {getByRole} = render(
      <DefaultTestWrapper>
        <Form />
      </DefaultTestWrapper>,
    );

    const btn = getByRole('button');
    fireEvent(btn, 'press');

    const storeActions = defaultMockStore.getActions();

    expect(storeActions.length).toBe(0);
  });

  it('dispatches request action after text input and btn pressed', () => {
    const {getByRole, getByPlaceholderText} = render(
      <DefaultTestWrapper>
        <Form />
      </DefaultTestWrapper>,
    );

    const input = getByPlaceholderText('Enter a name');
    fireEvent(input, 'changeText', 'newname');

    const btn = getByRole('button');
    fireEvent(btn, 'press');

    const storeActions = defaultMockStore.getActions();
    const lastAction = storeActions[storeActions.length - 1];

    expect(actions.getGender).toHaveBeenLastCalledWith('newname');
    expect(lastAction).toEqual(actions.getGenderRequest());
  });
});
