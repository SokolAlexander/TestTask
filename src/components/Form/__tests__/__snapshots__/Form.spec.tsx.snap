// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches snapshot with default state 1`] = `
Array [
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter a name"
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderColor": "#141513",
        "borderWidth": 1,
        "marginVertical": 10,
        "paddingHorizontal": 15,
        "paddingVertical": 10,
      }
    }
    value=""
  />,
  <View
    style={
      Object {
        "alignItems": "center",
        "minHeight": 30,
      }
    }
  />,
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#EDEDED",
        "borderColor": "#DDD",
        "borderWidth": 2,
        "justifyContent": "center",
        "marginHorizontal": 25,
        "marginTop": 10,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#141513",
          "fontSize": 18,
        }
      }
    >
      Get Gender
    </Text>
  </View>,
]
`;

exports[`Form matches snapshot with request error state 1`] = `
Array [
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter a name"
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderColor": "#141513",
        "borderWidth": 1,
        "marginVertical": 10,
        "paddingHorizontal": 15,
        "paddingVertical": 10,
      }
    }
    value=""
  />,
  <View
    style={
      Object {
        "alignItems": "center",
        "minHeight": 30,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "red",
        }
      }
    >
      Test error
    </Text>
  </View>,
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#EDEDED",
        "borderColor": "#DDD",
        "borderWidth": 2,
        "justifyContent": "center",
        "marginHorizontal": 25,
        "marginTop": 10,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#141513",
          "fontSize": 18,
        }
      }
    >
      Get Gender
    </Text>
  </View>,
]
`;

exports[`Form matches snapshot with request pending state 1`] = `
Array [
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter a name"
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderColor": "#141513",
        "borderWidth": 1,
        "marginVertical": 10,
        "paddingHorizontal": 15,
        "paddingVertical": 10,
      }
    }
    value=""
  />,
  <View
    style={
      Object {
        "alignItems": "center",
        "minHeight": 30,
      }
    }
  >
    <ActivityIndicator
      color="black"
    />
  </View>,
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#EDEDED",
        "borderColor": "#DDD",
        "borderWidth": 2,
        "justifyContent": "center",
        "marginHorizontal": 25,
        "marginTop": 10,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#141513",
          "fontSize": 18,
        }
      }
    >
      Get Gender
    </Text>
  </View>,
]
`;
